src/charm.py

```diff
@@ -62,30 +59,33 @@ def _on_config_changed(self, event):
             logger.info("Invalid port number, 22 is reserved for SSH")
             return
 
-        self._update_layer_and_restart(None)
+        self._update_layer_and_restart()
+
+    def _on_demo_server_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
+        self._update_layer_and_restart()
 
     def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
         """Event is fired when postgres database is created."""
-        self._update_layer_and_restart(None)
+        self._update_layer_and_restart()
 
-    def _update_layer_and_restart(self, event) -> None:
+    def _update_layer_and_restart(self) -> None:
         """Define and start a workload using the Pebble API.
 
         You'll need to specify the right entrypoint and environment
         configuration for your specific workload. Tip: you can see the
         standard entrypoint of an existing container using docker inspect
 
+        Learn more about interacting with Pebble at https://juju.is/docs/sdk/pebble
         Learn more about Pebble layers at https://github.com/canonical/pebble
         """
 
         # Learn more about statuses in the SDK docs:
         # https://juju.is/docs/sdk/constructs#heading--statuses
         self.unit.status = ops.MaintenanceStatus("Assembling Pebble layers")
-        new_layer = self._pebble_layer.to_dict()
         try:
             # Get the current pebble layer config
             services = self.container.get_plan().to_dict().get("services", {})
-            if services != new_layer["services"]:
+            if services != self._pebble_layer.to_dict().get("services", {}):
```

Comment:
For what it's worth, the code shouldn't be like this, where it gets the existing plan and uses that to decide whether to restart services. It should just set the layer in Pebble and tell Pebble to replan, and Pebble will figure out whether to restart anything or not.

However, fixing that is probably out of scope here, so this change is fine.